include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -1578,6 +1585,1157 @@ namespace Differentiation
+       * that exist for use with the tensor classes.
+       */
+      template <typename NumberType>
+      const NumberType &
+      get_tensor_entry(const NumberType &t, const unsigned int &index)
+      {
+        Assert(index == 0, ExcIndexRange(index, 0, 1));
+        (void)index;
+        return t;
+      }
+
+
+      /**
+       * Return a reference to the entry stored in the @p index'th unrolled
+       * component of the generic tensor @p t.
+       */
+      template <int rank,
+                int dim,
+                typename NumberType,
+                template <int, int, typename> class TensorType>
+      NumberType &
+      get_tensor_entry(TensorType<rank, dim, NumberType> &t,
+                       const unsigned int &               index)
+      {
+        // Where possible, get values using TableIndices
+        Assert(index < t.n_independent_components,
+               ExcIndexRange(index, 0, t.n_independent_components));
+        return t[get_table_indices(t, index)];
+      }
+
+
+      /**
+       * Return a reference to the entry stored in the @p index'th unrolled
+       * component of the rank-0 tensor @p t.
+       */
+      template <int dim,
+                typename NumberType,
+                template <int, int, typename> class TensorType>
+      NumberType &get_tensor_entry(TensorType<0, dim, NumberType> &t,
+                                   const unsigned int &            index)
+      {
+        Assert(index == 0, ExcIndexRange(index, 0, 1));
+        (void)index;
+        return t;
+      }
+
+
+      /**
+       * Return a reference to  @p t.
+       * This function exists to provide compatibility with similar functions
+       * that exist for use with the tensor classes.
+       */
+      template <typename NumberType>
+      NumberType &
+      get_tensor_entry(NumberType &t, const unsigned int &index)
+      {
+        Assert(index == 0, ExcIndexRange(index, 0, 1));
+        (void)index;
+        return t;
+      }
+
+    } // namespace internal
+
+
+
+    /**
+     * A base helper class that facilitates the evaluation of point-wise defined
+     * functions. As it offers a field-based interface, this class would
+     * typically be used to compute the derivatives of a constitutive law
+     * defined at a quadrature point; however, it may also be used in other
+     * contexts, such as to compute the linearization of a set of local
+     * nonlinear equations.
+     *
+     * This class is not complete for use since no derivative computations have
+     * been implemeneted. What can be computed and how the computed values are
+     * interpreted depends on the number of dependent variables to be
+     * considered.
+     *
+     * @author Jean-Paul Pelteret, 2016, 2017, 2018
+     */
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType = double>
+    class ADHelperPointLevelFunctionsBase
```

Comment:
This is the point-wise counterpart of the `ADHelperCellLevelBase` class. One just needs a way to convey that these are for computations at a continuum point, or gauss point, rather than for finite-element level calculations. I'll try to clarify this more in the documentation.

Reply:
Ah, "at the point level" or "at the level of points".